packages/vite/src/node/plugins/asset.ts

```diff
@@ -228,12 +228,78 @@ async function fileToBuiltUrl(
     const contentHash = getAssetHash(content)
     const { search, hash } = parseUrl(id)
     const postfix = (search || '') + (hash || '')
-    const basename = path.basename(file)
-    const ext = path.extname(basename)
-    const fileName = path.posix.join(
-      config.build.assetsDir,
-      `${basename.slice(0, -ext.length)}.${contentHash}${ext}`
-    )
+
+    // create fileName using rollupOptions.output.assetFileNames
+    let fileName: string
+    {
+      const basename = path.basename(file)
+
+      // placeholders for `assetFileNames`
+      // `hash` is slightly different from the rollup's one
+      const _extname = path.extname(basename)
+      const _ext = _extname.substr(1)
+      const _name = basename.slice(0, -_extname.length)
+      const _hash = contentHash
+
+      let assetFileNames: OutputOptions['assetFileNames']
+      const output = config.build?.rollupOptions?.output
+      // only the object format is currently considered here
+      if (output && !Array.isArray(output)) {
+        assetFileNames = output.assetFileNames
+      }
+      // defaults to '<assetsDir>/[name].[hash][extname]'
+      // slightly different from rollup's one ('assets/[name]-[hash][extname]')
+      if (assetFileNames == null) {
+        assetFileNames = path.posix.join(
+          config.build.assetsDir,
+          '[name].[hash][extname]'
+        )
+      }
+
+      switch (typeof assetFileNames) {
```

Comment:
This shouldn't be a `switch`. If `assetFileNames is a function, then it should be first applied and it could generate a pattern that needs to be replaced. The result could contain placeholders like `[name]`. Check out how they are doing it in rollup here https://github.com/rollup/rollup/blob/ce2592df6b369b68d61f58c2ef8bf4695421146a/src/utils/FileEmitter.ts#L40. So we first need to apply the function if there is one, and then always do the replacements.

Reply:
Oops, I misunderstood the behavior.